// @flagship.io/js-sdk/dist/chunks/browser/esm/C5Km5D0Q.js
/**
 * All events posted from iframe
 */
var MSG_NAME_FROM_IFRAME;
(function (MSG_NAME_FROM_IFRAME) {
    MSG_NAME_FROM_IFRAME["QaAssistantClose"] = "ABTASTY_QA_ASSISTANT_CLOSE";
    MSG_NAME_FROM_IFRAME["FsApplyForcedVariations"] = "FS_APPLY_FORCED_VARIATIONS";
    MSG_NAME_FROM_IFRAME["FsResetForcedVariations"] = "FS_RESET_FORCED_VARIATIONS";
    MSG_NAME_FROM_IFRAME["FsQaAssistantReady"] = "FS_QA_ASSISTANT_READY";
    MSG_NAME_FROM_IFRAME["MinimizeQaAssistantClose"] = "ABTASTY_QA_MINIMIZE_QA_ASSISTANT_CLOSE";
    MSG_NAME_FROM_IFRAME["FsTriggerRender"] = "FS_TRIGGER_RENDER";
    MSG_NAME_FROM_IFRAME["QaAssistantPlatformChoiceLoaded"] = "ABTASTY_QA_ASSISTANT_PLATFORM_CHOICE_LOADED";
})(MSG_NAME_FROM_IFRAME || (MSG_NAME_FROM_IFRAME = {}));
/**
 * All events posted to iframe
 */
var MSG_NAME_TO_IFRAME;
(function (MSG_NAME_TO_IFRAME) {
    MSG_NAME_TO_IFRAME["FsUpdateVisitorAllocatedVariation"] = "FS_UPDATE_VISITOR_ALLOCATED_VARIATION";
    MSG_NAME_TO_IFRAME["FsVisitorExposedVariation"] = "FS_VISITOR_EXPOSED_VARIATION";
    MSG_NAME_TO_IFRAME["FsHIT"] = "FS_HIT";
})(MSG_NAME_TO_IFRAME || (MSG_NAME_TO_IFRAME = {}));
var VisitorVariationUpdateParam;
(function (VisitorVariationUpdateParam) {
    VisitorVariationUpdateParam["NewNavigation"] = "newNavigation";
})(VisitorVariationUpdateParam || (VisitorVariationUpdateParam = {}));
var INTERNAL_EVENTS;
(function (INTERNAL_EVENTS) {
    INTERNAL_EVENTS["FsTriggerRendering"] = "FS_TRIGGER_RENDERING";
})(INTERNAL_EVENTS || (INTERNAL_EVENTS = {}));

function sendMessageToIframe(data) {
    if (!window?.frames?.ABTastyQaAssistant) {
        return;
    }
    window.frames.ABTastyQaAssistant.postMessage(data, '*');
}
function sendVisitorAllocatedVariations(visitorVariationState) {
    if (!visitorVariationState?.visitorVariations) {
        return;
    }
    sendMessageToIframe({
        name: MSG_NAME_TO_IFRAME.FsUpdateVisitorAllocatedVariation,
        value: visitorVariationState.visitorVariations
    });
}
function sendVisitorExposedVariations(visitorVariationState) {
    if (!visitorVariationState?.exposedVariations) {
        return;
    }
    const navigationDetected = visitorVariationState.navigationDetected;
    if (visitorVariationState.navigationDetected) {
        visitorVariationState.navigationDetected = false;
    }
    sendMessageToIframe({
        name: MSG_NAME_TO_IFRAME.FsVisitorExposedVariation,
        value: visitorVariationState.exposedVariations,
        param: navigationDetected ? VisitorVariationUpdateParam.NewNavigation : undefined
    });
}
function sendFsHitToQA(hit) {
    sendMessageToIframe({
        name: MSG_NAME_TO_IFRAME.FsHIT,
        value: hit.map(item => {
            return {
                ...item,
                timestamp: Date.now() - item.qt
            };
        })
    });
}

var index = /*#__PURE__*/Object.freeze({
    __proto__: null,
    sendFsHitToQA: sendFsHitToQA,
    sendMessageToIframe: sendMessageToIframe,
    sendVisitorAllocatedVariations: sendVisitorAllocatedVariations,
    sendVisitorExposedVariations: sendVisitorExposedVariations
});

export { INTERNAL_EVENTS as I, MSG_NAME_FROM_IFRAME as M, sendVisitorExposedVariations as a, index as i, sendVisitorAllocatedVariations as s };
